pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1482,17 +1493,50 @@ func (og *operationGenerator) verifyVolumeIsSafeToDetach(
 			"DetachVolume failed fetching node from API server",
 			fmt.Errorf("node object retrieved from API server is nil"))
 	}
+	if fetchErr != nil {
+		if errors.IsNotFound(fetchErr) {
+			klog.Warningf(volumeToDetach.GenerateMsgDetailed("Node not found on API server. DetachVolume will skip safe to detach check", ""))
+		} else {
+			// On failure, return error. Caller will log and retry.
+			return volumeToDetach.GenerateErrorDetailed("DetachVolume failed fetching node from API server", fetchErr)
+		}
+	}
 
-	for _, inUseVolume := range node.Status.VolumesInUse {
-		if inUseVolume == volumeToDetach.VolumeName {
-			return volumeToDetach.GenerateErrorDetailed(
-				"DetachVolume failed",
-				fmt.Errorf("volume is still in use by node, according to Node status"))
+	if needVerify {
+		for _, inUseVolume := range node.Status.VolumesInUse {
+			if inUseVolume == volumeToDetach.VolumeName {
+				return volumeToDetach.GenerateErrorDetailed(
+					"DetachVolume failed",
+					fmt.Errorf("volume is still in use by node, according to Node status"))
+			}
 		}
+		// Volume is not marked as in use by node
+		klog.Infof(volumeToDetach.GenerateMsgDetailed("Verified volume is safe to detach", ""))
 	}
 
-	// Volume is not marked as in use by node
-	klog.Infof(volumeToDetach.GenerateMsgDetailed("Verified volume is safe to detach", ""))
+	// Update node status to remove the volume from the attached list
+	err := actualStateOfWorld.RemoveVolumeFromReportAsAttached(
+		volumeToDetach.VolumeName, volumeToDetach.NodeName)
+	if err != nil {
+		klog.Warning(volumeToDetach.GenerateError("DetachVolume.Detach failed to remove volume from report as attached", err))
+	}
+	attachedVolumes := actualStateOfWorld.GetVolumesToReportAttachedForNode(volumeToDetach.NodeName)
+	nodeCopy := node.DeepCopy()
+	nodeCopy.Status.VolumesAttached = attachedVolumes
+	_, patchBytes, err := nodeutil.PatchNodeStatus(og.kubeClient.CoreV1(), volumeToDetach.NodeName, node, nodeCopy)
```

Comment:
This is problematic. Detach runs on a separate goroutine from the rest of the reconciler. With this, you are introducing a second thread which update the same field asynchronously with the primary reconciler loop. Let's discuss the implications. And can we keep node status updates only synchronously triggered in the loop?